packages/ember-routing/lib/location/history_location.js

```diff
@@ -189,17 +189,25 @@ export default EmberObject.extend({
     this._removeEventListener();
 
     this._popstateHandler = () => {
+      var url = this.getURL();
+      var shouldHandleURL = this._shouldHandleURL(url);
       // Ignore initial page load popstate event in Chrome
       if (!popstateFired) {
         popstateFired = true;
-        if (this.getURL() === this._previousURL) { return; }
+        if (!shouldHandleURL) { return; }
       }
-      callback(this.getURL());
+      if (!shouldHandleURL) { return; }
+      callback(url);
```

Comment:
I'm curious if there can just be one early return instead of two.

```js
    this._popstateHandler = () => {
      var url = this.getURL();
      if (!this._shouldHandleURL(url)) {
        return;
      }
      // Ignore initial page load popstate event in Chrome
      if (!popstateFired) {
        popstateFired = true;
      }
      callback(url);
    };
```